### DISCUSSION TITLE
Microservices with Deno as a Monorepo

### DISCUSSION THREAD

**Discussion by asjadanis** [denoland/deno#18143] [Q&A] [answered]
Is there any general guideline or recommended way of setting up a monorepo with Deno. I couldn't find much online resources on it. Since deno doesn't rely on a package manager or something like `package.json` how would one use existing monorepo solutions like turborepo as they mainly rely on `package.json`. Has anyone setup monorepos with Deno.

**Comment by jtoppine** [comment, marked as answer]
Correct me if I'm wrong, as I haven't actually used Turborepo. But I took a glance at some tutorials, and it appears to solve all kinds of problems that don't really have to exist in Deno.

So, for a monorepo, just have a folder with stuff, organized how you like it. Then for each project/microservice, you have an entrypoint file (just ts/js that imports whatever it needs). And you are basically done. 

Deno takes care of "compiling", refreshing, and caching when you run an entrypoint. Packaging a microservice for deployment/distribution is not so clearcut, but there is Deno.emit, esbuild and deno info command to help with that.

I might be missing something. I feel like I'm missing something, because this seems too simple. But then again, javascript tooling has become such a large, convoluted mess of build tools, frameworks, version managers and whatnots due to "history". And we are used to lean on these complex solutions, when something super basic and simple might work with Deno.

**Reply by bradenmacdonald** [reply]
Exactly. I have [a small monorepo](https://github.com/neolace-dev/neolace) for my project and it works really well without any specific tooling. It's one of the great things about Deno.

**Reply by carere** [reply]
@bradenmacdonald wonderful example, thx a lot for your share 😃

**Comment by danilo-valente** [comment]
Sometimes a simple folder with stuff inside might be enough, as @jtoppine mentioned, but in other situations you might have more complex use cases.

I couple months ago I stumbled across the same limitation and found this thread, but found no suitable tools for this issue, so I decided to create my own Deno monorepo manager.

## [Compy - Minimalist (yet helpful) monorepo manager for Deno](https://deno.land/x/compy)

https://deno.land/x/compy

> **Disclaimer:** Compy is still in early stage and some features might be redesigned in future versions. Automated testing and proper documentation is also in progress.

That said, it does provides good tooling for creating a project and its modules, adding dependencies and mapping them through import maps, running Deno's CLI commands within a module's context and linking modules between themselves. In fact, I'm currently using it in production.

Some commands include:
- `compy init <project>` to create a new project/workspace
- `compy mod <module>` to create a new module within the project
- `compy add std` to add the latest std version to all modules
- `compy add oak v12.6.1 -m <module>` to add `oak@v12.6.1` to an specific module
- `compy <cache|start|dev|run|test|fmt|lint> <module>` to run a command within a module's context
- and more

Here's a simple (outdated) demo:
![Untitled-2xc](https://github.com/denoland/deno/assets/1886960/e89d0c96-3a2a-43f7-98c7-d1d987554c55)

@bradenmacdonald @carere you might find this helpful as well.

**Comment by thoroc** [comment]
Have you tried `nx.dev` ? There is branch of their `deno` plugin: https://github.com/axhxrx/axhxrx-nx-deno.

I use `nx.dev` at work in conjunction with `AWS CDK` (albeit no deno `buy-in` for now) and it works surprisingly well.
